kernel/src/engine/default/json.rs

```diff
@@ -286,21 +557,172 @@ mod tests {
+            .map_ok(into_record_batch)
             .try_collect()
             .unwrap();
 
         assert_eq!(data.len(), 1);
         assert_eq!(data[0].num_rows(), 4);
+
+        // limit batch size
+        let handler = handler.with_batch_size(2);
+        let data: Vec<RecordBatch> = handler
+            .read_json_files(files, Arc::new(physical_schema.try_into().unwrap()), None)
+            .unwrap()
+            .map_ok(into_record_batch)
+            .try_collect()
+            .unwrap();
+
+        assert_eq!(data.len(), 2);
+        assert_eq!(data[0].num_rows(), 2);
+        assert_eq!(data[1].num_rows(), 2);
+    }
+
+    #[tokio::test]
+    async fn test_ordered_get_store() {
+        // note we don't want to go over 1000 since we only buffer 1000 requests at a time
+        let num_paths = 1000;
+        let ordered_paths: Vec<Path> = (0..num_paths)
+            .map(|i| Path::from(format!("/test/path{}", i)))
+            .collect();
+        let jumbled_paths: Vec<_> = ordered_paths[100..400]
+            .iter()
+            .chain(ordered_paths[400..].iter().rev())
+            .chain(ordered_paths[..100].iter())
+            .cloned()
+            .collect();
+
+        let memory_store = InMemory::new();
+        for (i, path) in ordered_paths.iter().enumerate() {
+            memory_store
+                .put(path, Bytes::from(format!("content_{}", i)).into())
+                .await
+                .unwrap();
+        }
+
+        // Create ordered store with natural order (0, 1, 2, ...)
+        let ordered_store = Arc::new(OrderedGetStore::new(
+            memory_store.fork(),
+            ordered_paths.clone(),
+        ));
+
+        let (tx, rx) = mpsc::channel();
+
+        // Spawn tasks to GET each path in our somewhat jumbled order
+        // They should complete in order (0, 1, 2, ...) due to OrderedGetStore
+        let handles = jumbled_paths.into_iter().map(|path| {
+            let store = ordered_store.clone();
+            let tx = tx.clone();
+            tokio::spawn(async move {
+                let _ = store.get(&path).await.unwrap();
+                tx.send(path).unwrap();
+            })
+        });
+
+        // TODO(zach): we need to join all the handles otherwise none of the tasks run? despite the
+        // docs?
+        future::join_all(handles).await;
+        drop(tx);
+
+        // NB (from mpsc::IntoIter): This iterator will block whenever next is called, waiting for
+        // a new message, and None will be returned if the corresponding channel has hung up.
+        let mut completed = Vec::new();
+        while let Ok(path) = rx.recv() {
+            completed.push(path);
+        }
+
+        assert_eq!(
+            completed,
+            ordered_paths.into_iter().collect_vec(),
+            "Expected paths to complete in order"
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 3)]
+    async fn test_read_json_files_ordering() {
```

Comment:
I think it might be helpful to have a test that exceeds the buffering limit.

Reply:
added!